### ISSUE TITLE
babel + child processes

### ISSUE THREAD

**Issue by boneskull** [wallabyjs/public#359]
Hi,

I'm unsure how to configure wallaby to handle a project in which a child process module, transpiled by Babel, is forked via `child_process.fork()`.

I have tried both `preprocessor` and `compile` style.  This is my config:

```js
'use strict';

var babel = require('babel-core');

module.exports = function wallabyConfig(wallaby) {
  return {
    files: [
      'src/**/*.js',
      {
        pattern: 'test/fixture.js',
        instrument: false
      },
      'package.json'
    ],
    tests: [
      'test/**/*.spec.js'
    ],
    env: {
      type: 'node',
      runner: 'node'
    },
    compilers: {
      '**/*.js': wallaby.compilers.babel({
        babel: babel,
        presets: ['es2015'],
        sourceMap: true
      })
    },
    testFramework: 'mocha',
    debug: true,
    bootstrap: function bootstrap(wallaby) {
      const path = require('path');
      // load transpiled fixture
      require(path.join(wallaby.projectCacheDir, 'test', 'fixture'));
    }
  };
};
```

I get:

```plain
wallaby.js started
core v1.0.149
Fri, 27 Nov 2015 01:03:11 GMT wallaby:project File cache: /Users/boneskull/Library/Caches/PyCharm50/wallaby/projects/a28f4494d731eb29
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Parallelism for initial run: 6, for regular run: 3
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #0
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #1
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #2
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #3
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #4
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Starting run worker instance #5
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Web server is listening at 57039
Fri, 27 Nov 2015 01:03:11 GMT wallaby:project Re-instrumenting original file because its ID has changed
Fri, 27 Nov 2015 01:03:11 GMT wallaby:project Preparing to process test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:11 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #0
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #1
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #2
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #4
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #5
Fri, 27 Nov 2015 01:03:11 GMT wallaby:workers Started run worker instance (delayed) #3
Fri, 27 Nov 2015 01:03:12 GMT wallaby:processor Running 1 processor(s) for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:14 GMT wallaby:processor Finished running one processor for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Processing compiled test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Instrumenting test/e2e/process-ipc.spec.js, via process pool: true
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project No preprocessors configured for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Writing to disk and caching processed file test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Stopping process pool
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project File cache is up-to-date, starting full test run
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Running postprocessor
Fri, 27 Nov 2015 01:03:14 GMT wallaby:postprocessor New TypeScript language service is required
Fri, 27 Nov 2015 01:03:14 GMT wallaby:postprocessor TypeScript postprocessor will be removed because in 5 project files none were found by patterns: **/*.ts
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Postprocessor execution finished
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Postprocessor is removed as requested by itself
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Test run started; run priority: 3
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Running all tests
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Starting test run, priority: 3
Fri, 27 Nov 2015 01:03:14 GMT wallaby:nodeRunner Starting sandbox [worker #0, session #q1i0u]
Fri, 27 Nov 2015 01:03:14 GMT wallaby:nodeRunner Preparing sandbox [worker #0, session #q1i0u]
Fri, 27 Nov 2015 01:03:14 GMT wallaby:nodeRunner Prepared sandbox [worker #0, session #q1i0u]
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Running tests in sandbox [worker #0, session #q1i0u]
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Loaded 1 test(s)
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Test executed: should fork a child process
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Run 1 test(s), skipped 0 test(s)
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Sandbox [q1i0u] is responsive, closing it
Finished executing 1 affected test(s)
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Test run finished
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Processed console.log entries
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Processed loading sequences
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Processed executed tests
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Processed code coverage
Fri, 27 Nov 2015 01:03:14 GMT wallaby:project Test run result processed and sent to IDE
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Sandbox (active) [q1i0u] error: /Users/boneskull/Library/Caches/PyCharm50/wallaby/projects/a28f4494d731eb29/instrumented/src/worker.js:2
Runtime error: ./src/worker.js:2
Fri, 27 Nov 2015 01:03:14 GMT wallaby:workers Sandbox [q1i0u] is responsive, closing it
$_$wp(3);
^

ReferenceError: $_$wp is not defined
    at Object.<anonymous> (/Users/boneskull/Library/Caches/PyCharm50/wallaby/projects/a28f4494d731eb29/instrumented/src/worker.js:2:1)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
    at node.js:963:3

```

I've also tried this configuration:

```js
'use strict';

var babel = require('babel-core');

module.exports = function wallabyConfig() {
  return {
    files: [
      'src/index.js',
      {
        pattern: 'test/fixture.js',
        instrument: false
      },
      {
        pattern: 'src/worker.js',
        instrument: false
      }
    ],
    tests: [
      'test/**/*.spec.js'
    ],
    env: {
      type: 'node',
      runner: 'node'
    },
    preprocessors: {
      '**/*.js': function (file) {
        return babel.transform(file.content, {
          sourceMap: true,
          presets: 'es2015'
        });
      }
    },
    testFramework: 'mocha',
    debug: true,
    bootstrap: function bootstrap(wallaby) {
      const path = require('path');
      require(path.join(wallaby.projectCacheDir, 'test', 'fixture'));
    }
  };
};
```

I thought perhaps that disabling instrumentation would help.  This doesn't throw the error, but rather complains that the `debug` module, which `worker.js` (forked by `index.js`) uses, is not found.

```plain
wallaby.js started
core v1.0.149
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File cache: /Users/boneskull/Library/Caches/PyCharm50/wallaby/projects/a28f4494d731eb29
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Config file change detected, invalidating local cache
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Parallelism for initial run: 6, for regular run: 3
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #0
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #1
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #2
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #3
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #4
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Starting run worker instance #5
Fri, 27 Nov 2015 01:32:53 GMT wallaby:workers Web server is listening at 59297
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File cache requires some updates, waiting required files from IDE
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File src/index.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File src/worker.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File test/fixture.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File test/e2e/process-ipc.spec.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File node_modules/debug/node.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File node_modules/lodash/index.js changes are patches only: false
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process src/index.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process src/worker.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File src/worker.js is not instrumented by core
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process test/fixture.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File test/fixture.js is not instrumented by core
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process node_modules/debug/node.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File node_modules/debug/node.js is not instrumented by core
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Preparing to process node_modules/lodash/index.js
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project Starting processor pool (if not yet started)
Fri, 27 Nov 2015 01:32:53 GMT wallaby:project File node_modules/lodash/index.js is not instrumented by core
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #0
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #1
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #2
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #3
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #4
Fri, 27 Nov 2015 01:32:54 GMT wallaby:workers Started run worker instance (delayed) #5
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor No compilers found for src/index.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:project Instrumenting src/index.js, via process pool: true
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor Running 1 processor(s) for test/fixture.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor Running 1 processor(s) for node_modules/debug/node.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor No compilers found for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:project Instrumenting test/e2e/process-ipc.spec.js, via process pool: true
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor Running 1 processor(s) for src/worker.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor Running 1 processor(s) for node_modules/lodash/index.js
Fri, 27 Nov 2015 01:32:55 GMT wallaby:processor Running 1 processor(s) for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:32:56 GMT wallaby:processor Running 1 processor(s) for src/index.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:processor Finished running one processor for test/fixture.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project Writing to disk and caching processed file test/fixture.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project File test/fixture.js changes saved, store ts: 1448585843000, change ts: 1448587973273
Fri, 27 Nov 2015 01:33:02 GMT wallaby:processor Finished running one processor for src/worker.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:processor Finished running one processor for test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project Writing to disk and caching processed file src/worker.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project Writing to disk and caching processed file test/e2e/process-ipc.spec.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project File src/worker.js changes saved, store ts: 1448585410000, change ts: 1448587973267
Fri, 27 Nov 2015 01:33:02 GMT wallaby:processor Finished running one processor for node_modules/debug/node.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project File test/e2e/process-ipc.spec.js changes saved, store ts: 1448587039000, change ts: 1448587973279
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project Writing to disk and caching processed file node_modules/debug/node.js
Fri, 27 Nov 2015 01:33:02 GMT wallaby:project File node_modules/debug/node.js changes saved, store ts: 1426272474000, change ts: 1448587973284
Fri, 27 Nov 2015 01:33:03 GMT wallaby:processor Finished running one processor for src/index.js
Fri, 27 Nov 2015 01:33:03 GMT wallaby:project Writing to disk and caching processed file src/index.js
Fri, 27 Nov 2015 01:33:03 GMT wallaby:project File src/index.js changes saved, store ts: 1448587422000, change ts: 1448587973253
[BABEL] Note: The code generator has deoptimised the styling of "unknown" as it exceeds the max of "100KB".
Fri, 27 Nov 2015 01:33:05 GMT wallaby:processor Finished running one processor for node_modules/lodash/index.js
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Writing to disk and caching processed file node_modules/lodash/index.js
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project File node_modules/lodash/index.js changes saved, store ts: 1438668224000, change ts: 1448587973290
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Stopping process pool
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Running postprocessor
Fri, 27 Nov 2015 01:33:05 GMT wallaby:postprocessor New TypeScript language service is required
Fri, 27 Nov 2015 01:33:05 GMT wallaby:postprocessor TypeScript postprocessor will be removed because in 6 project files none were found by patterns: **/*.ts
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Postprocessor execution finished
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Postprocessor is removed as requested by itself
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Test run started; run priority: 3
Fri, 27 Nov 2015 01:33:05 GMT wallaby:project Running all tests
Fri, 27 Nov 2015 01:33:05 GMT wallaby:workers Starting test run, priority: 3
Fri, 27 Nov 2015 01:33:05 GMT wallaby:nodeRunner Starting sandbox [worker #0, session #gwwq4]
Fri, 27 Nov 2015 01:33:05 GMT wallaby:nodeRunner Preparing sandbox [worker #0, session #gwwq4]
Fri, 27 Nov 2015 01:33:05 GMT wallaby:nodeRunner Prepared sandbox [worker #0, session #gwwq4]
Fri, 27 Nov 2015 01:33:05 GMT wallaby:workers Running tests in sandbox [worker #0, session #gwwq4]
Fri, 27 Nov 2015 01:33:06 GMT wallaby:workers Loaded 1 test(s)
Fri, 27 Nov 2015 01:33:06 GMT wallaby:workers Sandbox (active) [gwwq4] error: module.js:339
Runtime error: module.js:339
    throw err;
    ^

Error: Cannot find module 'debug'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (./src/worker.js:5:13)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)

Fri, 27 Nov 2015 01:33:06 GMT wallaby:workers Run 0 test(s), skipped 0 test(s)
Fri, 27 Nov 2015 01:33:06 GMT wallaby:workers Sandbox [gwwq4] is responsive, closing it
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Test run finished
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Processed console.log entries
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Processed loading sequences
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Processed executed tests
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Processed code coverage
Fri, 27 Nov 2015 01:33:06 GMT wallaby:project Test run result processed and sent to IDE
```

The `debug` module is indeed present.  I tried including it explicitly by adding it to the `files` array, but no dice.  Please show me where I've messed up.  I feel like I have asked this before. :smile:

**Comment by ArtemGovorov** [comment]
A combo of `instrument: false` for the worker + setting `NODE_PATH` in wallaby config should help:
``` JavaScript
module.exports = function (wallaby) {
  process.env.NODE_PATH += path.delimiter + path.join(wallaby.localProjectDir, 'node_modules');
  return {...}
}
```

If you'd like to get the code coverage for worker itself, you may want to test it separately by `require`-ing it directly and testing it in isolation.

**Comment by sindresorhus** [comment]
@ArtemGovorov You might want to look into using [`nyc`](https://github.com/bcoe/nyc) which works through subprocesses boundaries.

// @bcoe

**Comment by ArtemGovorov** [comment]
@sindresorhus Thanks! Will have a look. 

Would also be interesting to understand how frequently people do actually need multiprocess coverage. I personally would normally test a module in isolation by just `require`-ing it from tests (with coverage in a single process), so the tests are fast because of no process creation/communication. Then there may be a test or two to test the integration where I would spin up the module in subprocess from another module, but in this case I don't really need the coverage for the module in the subprocess, because I'm not testing the module itself, but just the interaction with it.

**Comment by sindresorhus** [comment]
@ArtemGovorov We use `nyc` with AVA, as AVA run test files in separate processes for the purpose of parallelism and clean environments. Seeing how JS is single-threaded, multiple processes is a good way to speed up heavy sync operations.

**Comment by bcoe** [comment]
@ArtemGovorov in my travels with I've definitely found that a good chunk of Node.js testing frameworks, e.g., `AVA`, `node-tap`, opt to isolate tests in subprocesses.

Let me know if you find any of the logic in `nyc` useful, I'd be happy to split it into more modules, depending on the bits and pieces of functionality that other library maintainers are itching to take advantage of.

**Comment by ArtemGovorov** [comment]
@sindresorhus @bcoe Thanks guys. Supporting parallel runs of `AVA` tests in wallaby shouldn't be an issue. Wallaby runs test files in parallel for other testing frameworks as well and has the mechanism to merge the coverage. 

So in our integration wallaby basically acts as a **runner** (it knows what to run and when), while `AVA` (and other frameworks that we support) mostly acts as a **testing framework** (it provides what to run, the API, test structure, assertions mechanism, etc.). I know `AVA` (and pretty much all other testing frameworks) provide a test runner as well. We re-use those bits that we can from them in wallaby, and adding those that are missing. Also trying to keep testing framework/runner unique features - for example in `AVA` integration we'll try to use its awesome feature of non-blocking async tests (executing in non serial manner, without waiting other tests to finish).

I was actually wondering how frequently users are writing tests that (themselves or the code they test) are spinning up subprocesses **and** need the cross process coverage. 

For example the original issue was that some code/test does `child_process.fork()` and code coverage in the forked file is expected. As I have said I'd have separate tests (test files) for testing the forked module and the original module (where I'd have the process forking mocked in the most of the tests, except one or two to test the integration and don't really need the coverage of the forked module for the integration tests).

**Comment by ArtemGovorov** [comment]
@boneskull Let me know if [the solution worked for you](https://github.com/wallabyjs/public/issues/359#issuecomment-160021420).

**Comment by boneskull** [comment]
@ArtemGovorov well, I looked at this for about five years, but am still struggling with child processes (in mocha, of course).  Here's the wallaby config as it looks in my working copy (of which you wrote much of yourself):

```js
'use strict';

module.exports = () => {
  return {
    files: [
      'index.js',
      'lib/**/*.{js,json}',
      'test/setup.js',
      'test/assertions.js',
      {
        pattern: 'bin/*', 
        instrument: false
      },
      {
        pattern: 'test/node-unit/**/*.fixture.js',
        instrument: false
      },
      {
        pattern: 'test/unit/**/*.fixture.js',
        instrument: false
      },
      {
        pattern: 'test/integration/fixtures/**/*.fixture.js',
        instrument: false
      },
      {
        pattern: 'test/integration/helpers.js',
        instrument: false
      },
      'package.json',
      'test/opts/mocha.opts',
      'mocharc.yml'
    ],
    filesWithNoCoverageCalculated: [
      'bin/*',
      'test/**/*.fixture.js',
      'test/setup.js',
      'test/assertions.js',
      'test/integration/helpers.js',
      'lib/browser/**/*.js'
    ],
    tests: [
      'test/unit/**/*.spec.js',
      'test/node-unit/**/*.spec.js',
      'test/integration/plugins/global-fixtures.spec.js'
    ],
    env: {
      type: 'node',
      runner: 'node'
    },
    workers: {recycle: true},
    testFramework: {type: 'mocha', path: __dirname},
    setup(wallaby) {
      const path = require('path');
      process.env.NODE_PATH += `${path.delimiter}${path.join(
        wallaby.localProjectDir,
        'node_modules'
      )}`;
      // running mocha instance is not the same as mocha under test,
      // running mocha is the project's source code mocha, mocha under test is instrumented version of the source code
      const runningMocha = wallaby.testFramework;
      runningMocha.timeout(1000);
      // to expose it/describe etc. on the mocha under test
      const MochaUnderTest = require('./');
      const mochaUnderTest = new MochaUnderTest();
      mochaUnderTest.suite.emit(
        MochaUnderTest.Suite.constants.EVENT_FILE_PRE_REQUIRE,
        global,
        '',
        mochaUnderTest
      );
      require('./test/setup');
    },
    debug: true,
    preservePaths: true
  };
};
```

Previously, we were running only _unit_ tests with Wallaby, but I thought I'd give an integration test a try.  The test in question is `test/integration/plugins/global-fixtures.spec.js`.  It uses `test/integration/helpers.js` to spawn `bin/mocha` against a fixture script in `test/integration/fixtures/somewhere`.

I added the `NODE_PATH` and added `instrument: false` for `bin/*` (for `bin/mocha`).  In this case, `lib/utils.js` required from `bin/mocha` throws:

```
/Users/boneskull/.vscode/extensions/wallabyjs.wallaby-vscode-1.0.229/projects/bf8cd1e3e23021b8/instrumented/lib/utils.js:2 
var $_$c = $_$wp(17); 
           ^ 
 
ReferenceError: $_$wp is not defined 
  at ​lib/utils.js:1​
  at Object.<anonymous> ​lib/utils.js:1​
  at Object.<anonymous> ​bin/mocha:13​
```

If I remove the `instrumented: false` from `bin/*`, I get the same error, but from `bin/mocha` instead:

```
/Users/boneskull/.vscode/extensions/wallabyjs.wallaby-vscode-1.0.229/projects/bf8cd1e3e23021b8/instrumented/bin/mocha:2 
var $_$c = $_$wp(108); 
           ^ 
 
ReferenceError: $_$wp is not defined 
  at ​bin/mocha:1​
  at Object.<anonymous> ​bin/mocha:1​
```

So I'm guessing that there's just a global variable `$_$wp` that's missing in the environment of the spawned `bin/mocha` process.  Is this possible to get working, or should I forget about it?  FWIW, the `NODE_PATH` thing has no effect.

**Comment by smcenlly** [comment]
Are you able to share the version of the codebase that has `test/integration/plugins/global-fixtures.spec.js`? We took a look at `master` branch and that file is missing. We'd like to take a see what you're doing to see how you might work around the issue.

**Comment by boneskull** [comment]
Thanks.  this branch: https://github.com/mochajs/mocha/tree/boneskull/issue/4308-global-setup-teardown

There's a bit of indirection here, as we have test helpers to spawn subprocesses.  [This](https://github.com/mochajs/mocha/blob/5ec47835e61e85a94b4b13239a3ec369469cffdd/test/integration/helpers.js#L305) is the function which will ultimately spawn the subprocess.

(the `.wallaby.js` file doesn't contain the changes above though, because it's broken)

**Comment by smcenlly** [comment]
Thanks for sending through the branch. You're seeing these errors because the Wallaby runtime doesn't exist when you're running the instrumented Wallaby code in the spawned processes outside of Wallaby's runtime environment.

We recommend adding a preprocessor for your `bin/mocha.js` entry point to "no-op" the Wallaby instrumentation calls when `bin/mocha.js` is used as the entry point for your sub-process:

```
...
    preprocessors: {
      "bin/*": f => {
        return `global.$_$wpe = global.$_$wp = ` +
        `global.$_$wf = global.$_$w = global.$_$wv = () => {};` +
        ` global.$_$tracer = { log: () => {} };//${f.content}`
      }
    },
...
```

After adding the preprocessor, we received another issue:

```
internal/validators.js:121 
    throw new ERR_INVALID_ARG_TYPE(name, 'string', value); 
    ^ 
 
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received an instance of Array 
 
  at exports.findConfig ​lib/cli/config.js:97:2
```

One thing to note is that Wallaby will not function on code in your sub-process (e.g. no code-coverage, no logging, time travel debugger, etc).

It looks like this might be expected and we figured was unrelated to Wallaby. Let us know if you need any further assistance.

**Comment by boneskull** [comment]
FWIW, this works.  I am a little concerned that it might be brittle, as it relies on stubbing an undocumented/private API. 
 Perhaps Wallaby could provide a public method that would configure the stub?  I think I'm asking for a built-in preprocessor.

**Comment by smcenlly** [comment]
You're right. The approach is brittle and you are stubbing undocumented/private API methods. Also, I don't think it's right for us to expose this setup code via an API or support it against breaking changes.

Thinking a little more about your problem, while what I suggested previously works, I don't think it's the right/best approach. Since Wallaby is not going to be reporting on any of your coverage with this approach, I would run mocha on your actual project files instead of the version of the files that are processed by Wallaby. I'm assuming the results will be the same either way. What I would do is:

1. In your `.wallaby.js` configuration file, update your setup file to set a `process.env` variable that you can use in your tests that captures the real project directory:

```diff
'use strict';

module.exports = () => {
  return {
    ...
    setup(wallaby) {
+     process.env.PROJECT_ROOT_DIR = wallaby.localProjectDir;
      ...
    },
    ...
  };
  ...
};
```

2. Update your code that spawns external processes to switch on whether `process.env.PROJECT_ROOT_DIR` is available. For example:

**test\integration\helpers.js**
```diff
/**
 * Path to `mocha` executable
 */
- const MOCHA_EXECUTABLE = require.resolve('../../bin/mocha');
+ const MOCHA_EXECUTABLE = process.env.PROJECT_ROOT_DIR ? path.join(process.env.PROJECT_ROOT_DIR, 'bin', 'mocha') : require.resolve('../../bin/mocha');
```

There are definitely some other changes you would need to make (e.g. `resolveFixturePath`), but this gives you the idea.

The only downside with the approach is that your code/tests will have some artifacts that are specific for when Wallaby is running but I still think the approach is reasonable and should be identical to when you're spawning your tests outside of Wallaby (i.e. when using mocha from the CLI).

**Comment by boneskull** [comment]
when running Mocha as a child process, against test fixture files, I don't care about reporting coverage on those files. (they are already excluded in the config)

I would like to report coverage on mocha internals in that subprocess, though. so if I use mocha flag _x_ in my subprocess, I want to see that the coverage for flag _x_'s code path. 

would what you are proposing inhibit that?

**Comment by smcenlly** [comment]
Neither approach supports what you are describing. Because you are launching Mocha in a sub-process, Wallaby's communication mechanism for reporting test results is not set up which is why we had to exclude `.bin` folder from instrumentation in the first approach. It is also why we were patching those internal methods. Both approaches will report on your code up until launching the sub-processes.

Wallaby does not currently support running/reporting on your tests/code from spawned processes. This would be a pretty big architectural change and not something that I think we can easily add as a supported behavior. I'll have a chat with the rest of the team later today about whether we might be able to support this.

Perhaps not ideal in terms of isolating from mocha internal state changes, but is it possible to launch mocha (e.g. with `require`) (instead of with a sub-process) for these tests?

**Comment by boneskull** [comment]
starting them w `require` would rather defeat the purpose, unfortunately. 

these are end-to-end tests, so, really they seem somewhat outside of wallaby's scope as I understand it. it is not fast to rerun these tests and I have disabled run-on-keystroke due to it. the code coverage reporting in subprocesses is helpful, but not critical--we get code coverage w/these tests in our CI.

at any rate, I will try the alternative you presented above, since it sounds like it would address my anxiety around depending on wallaby internals. 🙂 

thanks for helping out!
